Truck Inspection Plugin Client

qswawrq opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description
Address #613. This PR is still in progress until we have a complete definition of truck safety message.
<!--- Describe your changes in detail -->
## Related Issue
#613
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
Implement new feature.
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
Not tested.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

maefromm commented:
@qswawrq not sure if you got the chance to ask John, but I mentioned about potentially adding the logic that we can pull from the platform, instead of just the yaml file. @TonysCousin should add the logic to pull in the values dynamically instead of hard-coding in a yaml file? I think some of these values could be added/pulled from the https://github.com/usdot-fhwa-stol/CARMAConfig/blob/develop/freightliner_cascadia_2012/VehicleConfigParams.yaml dynamically.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,33 @@
+#50-74 (Yellow font) (Optional- Inspection Warranted – Next Level Priority)
+#75-100 (Red font) - (Inspect – Inspection warranted – top priority)
+iss_score: 49
+
```
missing 2 fields ads_health_status and ads_pretrip_check. NOTE I will send you the latest on naming convention so more to come. Awaiting client input.

maefromm commented on truck_inspection_client/src/truck_inspection_client.cpp:
```diff
@@ -0,0 +1,61 @@
+        pnh_->getParam("date_of_last_ads_calibration", date_of_last_ads_calibration_);
+        pnh_->getParam("iss_score", iss_score_);
+        pnh_->getParam("permit_required", permit_required_);
+        mo_pub_ = nh_->advertise<cav_msgs::MobilityOperation>("mobility_operation_outbound", 5);
```
missing 2 more fields, see comments on yaml.

TonysCousin commented:
I'd like to see ads_software_version dynamically assigned from the actual operating version ID (from the docker image).  ADS health status should be dynamically derived from health monitor info to reflect actual status of each driver (if not all green we need to come up with some logic to report what is going on).  Otherwise, I'm okay having all these come from the param file as is; if we pull them from the truck config, then it seems we need some logic to determine if it's a truck or not, because some of these params will not apply to passenger cars; not worth that effort right now.

maefromm commented:
@TonysCousin I'm good with that. But if this becomes another future demo, we should make it pull from truck config and include the weight in there. 

 I think for the UI screen, i'm surprised we didn't include the vehicle year, make and model "Freightliner" "Cascadia" "2012", which is from the truck config. Anyways, I haven't heard any additional inputs, so I think we're good for now. I will followup today.

qswawrq commented:
> @TonysCousin I'm good with that. But if this becomes another future demo, we should make it pull from truck config and include the weight in there.
> 
> I think for the UI screen, i'm surprised we didn't include the vehicle year, make and model "Freightliner" "Cascadia" "2012", which is from the truck config. Anyways, I haven't heard any additional inputs, so I think we're good for now. I will followup today.

The main reason I did not put those params in CARMAConfig file is the difficulty on deciding which should stay here and which should go there. The idea of CARMAConfig file is to hold vehicle specific  params. Following that definition, most of parameters in truck inspection plugin are vehicle specific and should be put in CARMAConfig. But they are also used by this single plugin in the near future. So I decide to keep them here.

maefromm commented:
@qswawrq yes agree. And with the different truck weight and trailer weight, not sure if that should be in the one CARMAConfig that we currently have for cascadia. So please proceed with the changes John suggested above and missing fields i mentioned.

TonysCousin commented:
Yes, weight, at least, would better come from vehicle config.  If that field doesn't already exist and you were to add that field, would that require changes in other code?  I doubt it, as it's just a param they don't have to read, right?  I'd still prefer truck-specific info to stay where it is for this iteration, just for simplicity.

maefromm commented:
@TonysCousin Shuwei's asking if the health status (besides health monitor and/or system_alert), should that also include the automated status of the CARMA vehicle (engage/disengaged)? I think we discussed and we said that could be lumped into health status. Right now our UI, we remove "ENGAGED" on the description or Safety Log.

TonysCousin commented:
My feeling is that ENGAGED is implied, since these messages won't be available if the vehicle is in manual operation mode, right?  If that's not true, then we definitely need to include the engagement state in the health status somehow.

qswawrq commented:
> My feeling is that ENGAGED is implied, since these messages won't be available if the vehicle is in manual operation mode, right? If that's not true, then we definitely need to include the engagement state in the health status somehow.

Are you saying to only send out this message when the platform is engaged?

maefromm commented:
@TonysCousin we can always listen to guidance state topic, but not sure if you are more interested if it's ENABLED vs. ENGAGED/ACTIVE/INACTIVE/and other guidance states. Are you just looking to see if the ADS system has been enabled but not actually driving fully automated?

TonysCousin commented:
Hmm, I'd have to refresh my understanding of that state model.  I'm thinking from a simplistic end user point of view for now.  That user will only care if the truck is actively being controlled by the software or by a human.  Subtle differences of engaged vs enabled vs active, etc, will be totally confusing.  We need it to be VERY SIMPLE:  currently AUTOMATED or HUMAN OPERATOR only.  If this AUTOMATED "state" can be determined simply by the receipt of messages, so much the better (i.e. if it is under human control, will the Message node still interact with the police car's request?)

maefromm commented:
@qswawrq can answer your message node question. 
If his answer is no, then the only way we know vehicle is AUTOMATED (not HUMAN), is to look at guidance state = ENGAGE i think.

qswawrq commented:
> Hmm, I'd have to refresh my understanding of that state model. I'm thinking from a simplistic end user point of view for now. That user will only care if the truck is actively being controlled by the software or by a human. Subtle differences of engaged vs enabled vs active, etc, will be totally confusing. We need it to be VERY SIMPLE: currently AUTOMATED or HUMAN OPERATOR only. If this AUTOMATED "state" can be determined simply by the receipt of messages, so much the better (i.e. if it is under human control, will the Message node still interact with the police car's request?)

Message node will keep running despite the vehicle state. For now, I used GREEN state to represent ENGAGE, and RED represents other states.

qswawrq commented:
@maefromm I think I addressed most of your comments. I intentionally kept the weight info here because I think it is a trip specific info (depends on the load of truck each trip?), which is not quite belong to CARMAConfig.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,37 @@
+# License plate number - 8 alphanumeric char
+license_plate: DOT-10003
+
+# Carrier Name - 20 alphanumeric char
```
Please update to 25 characters, since what they have now is more than 20.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,37 @@
+# Vehicle VIN number - 17 alphanumeric char
+vin_number: 1FUJGHDV0CLBP8834
+
+# License plate number - 8 alphanumeric char
```
Please update to 9 characters based on what they have now.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,37 @@
+weight: 100000
+
+# ADS Software Version: CARMA Platform v#.#.# (25characters)
+ads_software_version: CARMA Platform v3.3.0
```
Per John, he wants this from the CARMAConfig version like the UI

qswawrq replied:
Looks like that is something not doable from C++ node.

qswawrq replied:
Fixed as discussed.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,34 @@
+# Vehicle VIN number - 17 alphanumeric char
+vin_number: 1FUJGHDV0CLBP8834
```
Please update to "1FUJGHDV0CLBP8896" for Red Truck. The one above was for White Truck by mistake.

maefromm commented:
@qswawrq as discussed with @TonysCousin :

1. All of the headers of the topics you created for carma-platform and carma-messenger topics (if they have it), should have a timestamp.
2. The CARMAPlatform mobilityoperation topics shall have the header's sender_ bsm_id.
3. The remaining fields on the headers can be left blank.
4. With the timestamp on the header, the CarmaPlatform truck inspection client does not need to have it on the strategy_params. However carma-messenger truck_safety_info will still need that field, so carma-messenger truck inspection plugin will need to digest it and add it back.

Let me know if I had missed anything.

maefromm commented:
@qswawrq as discussed, yaml changes requested below to match red truck specs that will be used as the primary demo truck. I am awaiting monica's input on the weight and confirmation of the plate #s.

maefromm commented on truck_inspection_client/config/parameters.yaml:
```diff
@@ -0,0 +1,34 @@
+carrier_id: USDOT 1
+
+# Weight: ###,### lbs. (6 number format)
+weight: 100000
```
16639
